First commits

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR contains the initial code for this project, which is a Terraform module for creating an IAM role and policy that can read Terraform state objects from a specified S3 bucket.  It also creates a policy that allows the role to be assumed from a specified list of AWS account IDs.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

In [this comment](https://github.com/cisagov/cool-userservices-networking/pull/5#pullrequestreview-602926711), @jsf9k suggested the idea for this module, which would DRY out code from at least three other repositories- see https://github.com/cisagov/cool-system/issues/177 for details and justification.

Closes https://github.com/cisagov/cool-system/issues/177.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

Using the code in the `examples/basic_usage` directory, I tested this module by creating a series of test roles that could read the Terraform state for the example, varying the Terraform workspace each time.  I created roles that were allowed to access the following workspaces for the Terraform state:
* `default`
* `foobar`
*  `*` (any workspace)

I verified that the test roles I created:
* Were able to be assumed only by the Users account and no others
* Only allowed read access to the specified Terraform state resources and no others

All pre-commit automated checks pass.

<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
* [x] All new and existing tests pass.

dav3r committed:
Remove unneeded files inherited from the skeleton

dav3r committed:
Define providers

dav3r committed:
Define all necessary variables

dav3r committed:
Create some useful local variables

dav3r committed:
Create policy and role to read the specified Terraform state from the specified bucket

dav3r committed:
Create a policy allowing assumption of the role that can read the specified Terraform state

dav3r committed:
Define relevant outputs and remove those inherited from skeleton

dav3r committed:
Create a useful example

dav3r committed:
Update with useful information and remove skeleton text

dav3r committed:
Satisfy markdownlint pre-commit hook

dav3r committed:
Improve role description and make it more accurate

dav3r committed:
Properly handle permissions for default vs. non-default workspaces

felddy commented on README.md:
```diff
@@ -47,40 +40,38 @@ module "example" {
-| subnet_id | The ID of the AWS subnet to deploy into (e.g. subnet-0123456789abcdef0) | `string` | n/a | yes |
-| tags | Tags to apply to all AWS resources created | `map(string)` | `{}` | no |
+| account_ids | AWS account IDs that are allowed to assume the role that allows read-only access to the specified Terraform state. | `list(string)` | `[]` | no |
+| assume_role_policy_description | The description to associate with the IAM policy that allows assumption of the role that allows read-only access to the specified Terraform state.  Note that the first "%s" in this value will get replaced with the role_name variable and the second "%s" will get replaced with the terraform_account_name variable. | `string` | `Allow assumption of the %s role in th ...
```
Nifty

felddy commented on README.md:
```diff
@@ -2,38 +2,31 @@
+
+  account_ids = ["111111111111"]
+  role_name = "ReadTerraformStateReadRoleTFModuleTerraformState"
+  terraform_state_bucket_name = "cisa-cool-terraform-state"
```
Maybe make this match the tfvar example:

```suggestion
  terraform_state_bucket_name = "example-terraform-state-bucket"
```

dav3r replied:
I think we decided long ago that the bucket name is not particularly sensitive.  If that were to change, we would have [a fair amount of files to change](https://github.com/search?q=org%3Acisagov+cisa-cool-terraform-state&type=code).

I'm going to keep the bucket name as is, unless you strongly disagree with that plan.

felddy commented on examples/basic_usage/backend.tf:
```diff
@@ -0,0 +1,10 @@
+terraform {
+  backend "s3" {
+    bucket         = "cisa-cool-terraform-state"
```
Same suggestion... make this match the tfvar example:

```suggestion
    bucket         = "example-terraform-state-bucket"
```

dav3r replied:
See comment above.  I'm going to keep the bucket name as is, unless you strongly disagree with that plan.

felddy commented on examples/basic_usage/main.tf:
```diff
@@ -13,11 +17,12 @@ provider "aws" {
-  subnet_id             = aws_subnet.example.id
+  account_ids                 = var.account_ids
+  role_name                   = "ReadTerraformStateReadRoleTFModuleTerraformState"
+  terraform_state_bucket_name = "cisa-cool-terraform-state"
```
```suggestion
  terraform_state_bucket_name = "example-terraform-state-bucket"
```

dav3r replied:
See comment above.  I'm going to keep the bucket name as is, unless you strongly disagree with that plan.

felddy commented on read_terraform_state_policy.tf:
```diff
@@ -0,0 +1,39 @@
+    resources = concat(
+      # If the "default" workspace (or "*" for all workspaces) is specified,
+      # include the path where Terraform stores default workspace state.
+      contains(["*", "default"], var.terraform_workspace) ? ["arn:aws:s3:::${var.terraform_state_bucket_name}/${var.terraform_state_path}"] : [],
+      # For non-"default" workspaces, include the correct bucket path.
+      var.terraform_workspace != "default" ? ["arn:aws:s3:::${var.terraform_state_bucket_name}/env:/${var.terraform_workspace}/${var.terraform_state_path}"] : [],
```
This all comes from the backend configuration?  Just making sure it's not fragile.
https://www.terraform.io/docs/language/settings/backends/s3.html#key

dav3r replied:
It's a bit fragile- if Terraform were to change the path(s) where it stores state objects in S3, this would break.  However I'm not sure how to get around that.

felddy reviewed (approved):
Hey looks who's doing a review.  This guy. 👍 
What I like best about this PR is the lack of `venom` references.  Bravo.  
I had a question about s3 key paths, but I think I understand it. 
Also suggested example bucket name normalization.  
Strong work!  💪

jsf9k reviewed (approved):
Approved, but I feel like there is a lot of common jazz with [cisagov/s3-read-role-tf-module](https://github.com/cisagov/s3-read-role-tf-module).  Should we at least create an issue to DRY out these two Terraform modules?  Both create roles that create read-only permissions to S3 bucket objects.

hillaryj reviewed (approved):
🌮

dav3r commented:
> Approved, but I feel like there is a lot of common jazz with [cisagov/s3-read-role-tf-module](https://github.com/cisagov/s3-read-role-tf-module). Should we at least create an issue to DRY out these two Terraform modules? Both create roles that create read-only permissions to S3 bucket objects.

Brilliant idea!  I am going to just have this repo call [cisagov/s3-read-role-tf-module](https://github.com/cisagov/s3-read-role-tf-module) to create the role and policy.  That's why you're the 🧠 🧠 🧠  behind this operation!

dav3r committed:
Ensure default provider has a region

dav3r committed:
Add bucket_paths_allowed_to_read local variable

dav3r committed:
Use cisagov/s3-read-role-tf-module to create the role and policy

This is cleaner than doing it ourselves.  Thanks @jsf9k for this 
excellent suggestion!

dav3r commented:
FYI, I re-requested reviews since I implemented @jsf9k's suggestion.

mcdonnnj commented on examples/basic_usage/README.md:
````diff
@@ -1,16 +1,29 @@
+  the account ID(s) that are allowed to assume the Terraform role that will
+  be created:
+
+  ```console
````
```suggestion
  ```hcl
```

mcdonnnj reviewed (approved):
Makes sense. Thanks for making this a module!

dav3r committed:
Prefer hcl to console formatting for example tfvars content

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>
